src/h5web/dataset-visualizer/DimensionMapper.tsx

```diff
@@ -1,86 +1,78 @@
-import React from 'react';
+import React, { Fragment } from 'react';
 import ReactSlider from 'react-slider';
+import { isNumber } from 'lodash-es';
 import { Vis, DimensionMapping } from './models';
-import { useVisProps } from './VisProvider';
 import ButtonGroup from '../visualizations/shared/ButtonGroup';
 import styles from './DimensionMapper.module.css';
 
 interface Props {
   activeVis: Vis;
+  rawDims: number[];
+  mapperState: DimensionMapping;
   onChange: (d: DimensionMapping) => void;
 }
 
 function DimensionMapper(props: Props): JSX.Element {
-  const { activeVis, onChange } = props;
-  const { rawDims, mapping } = useVisProps();
+  const { activeVis, rawDims, mapperState, onChange } = props;
 
   if (activeVis === Vis.Line && rawDims.length === 2) {
     return (
       <div className={styles.mapper}>
         <div className={styles.buttonGroupWrapper}>
-          <span className={styles.buttonGroupLabel}>X</span>
+          <span className={styles.dimensionName}>x</span>
           <ButtonGroup
             ariaLabel="Dimension as X"
             className={styles.buttonGroup}
             buttonClassName={styles.btn}
             buttons={Object.keys(rawDims).map(dimKey => {
               const dimIndex = Number(dimKey);
               return {
-                label: dimKey,
-                isSelected: mapping.x === dimIndex,
+                label: `D${dimIndex}`,
+                isSelected: mapperState[dimIndex] === 'x',
                 onClick: () => {
-                  if (mapping.x !== dimIndex) {
-                    const newMapping: DimensionMapping = {
-                      x: dimIndex,
-                      slicingIndices: { [mapping.x]: 0 },
-                    };
-                    onChange(newMapping);
+                  const prevX = mapperState.indexOf('x');
+                  if (prevX !== dimIndex) {
+                    const newMapperState = mapperState.slice();
+                    newMapperState[prevX] = 0;
+                    newMapperState[dimIndex] = 'x';
+                    onChange(newMapperState);
```

Comment:
With the new mapping `mapperState`, the change in the `x` dimension resulting in switching the position of `x` with the previous one and setting the slicing index of the previous dimension to 0.